File: mod/Nova.Core/Resources/Resource.cs
using System;
using System.Collections.Generic;

namespace Nova.Core.Resources;

// Resource domain controls which solver owns this resource's flow.
//
//   Topological — flows along vessel topology (pipes, decouplers, up-only
//                 edges). Stored in tanks, drained by drain priority,
//                 staged-aware. Examples: RP-1, LOX, LH2, Hydrazine.
//                 Solved by StagingFlowSystem (water-fill).
//
//   Uniform     — single vessel-wide pool, no topology distinctions.
//                 May be cyclic (closed-loop life support, ISRU recycle).
//                 Examples: ElectricCharge today; O2/CO2/H2O/heat
//                 tomorrow. Solved by ProcessFlowSystem (LP).
//
// Resources don't switch domains. The Accumulator pattern bridges
// when a component needs to consume one domain's resource but produce
// in the other (e.g. fuel cell consumes Topological LH2/LOX, produces
// Uniform EC).
public enum ResourceDomain {
  Topological,
  Uniform,
}

public class Resource {
  public string Name;
  public string Abbreviation;
  public UnitDefinition Unit;
  public double Density; // kg/U
  public ResourceDomain Domain;

  // Cryogenic boiloff baseline — fraction of slice capacity lost per
  // Earth day under stock MLI insulation. Anchored to real spacecraft
  // cryo data (LH₂ ~3%/day, LOX ~1%/day, CH₄ ~0.5%/day at typical
  // tank surface-to-volume ratios). Storable / non-cryogenic resources
  // are 0 — no boiloff, no tier hardware needed. Used by TankVolume
  // together with InsulationTierTable to compute net boiloff per slice.
  public double MliBoiloffFractionPerDay;

  // Operating temperature (Kelvin) of the fluid in a tank — its boiling
  // point at the storage pressure. Drives the cryocooler's required
  // ΔT (against AmbientK) and Carnot COP, so LH₂ at 20 K costs
  // dramatically more EC per watt of cooling than LOX at 90 K under
  // the same insulation tier. 0 for storables / non-cryogenic.
  public double BoilingPointK;

  // Latent heat of vaporization in J/kg at the storage temperature.
  // Closes the loop between "fraction of tank lost per day" and the
  // actual heat-leak wattage the cooler must remove:
  //   Q_leak (W) = capacity × frac/day × density × Lv / 86400
  // 0 for storables / non-cryogenic.
  public double LatentHeatJPerKg;

  private static Dictionary<string, Resource> registry = new();

  public class UnitDefinition {
    public string Name;
    public string Symbol;

    public static UnitDefinition Watt = new() {
      Name = "Watt",
      Symbol = "W",
    };

    public static UnitDefinition Liter = new() {
      Name = "Liter",
      Symbol = "L",
    };
  }

  static Resource() {
    // Convention: EC quantity is Joules (1 EC = 1 J), EC rate is Watts
    // (1 EC/s = 1 W). The "Watt" label below is the canonical *rate*
    // unit; Buffer.Capacity / Buffer.Contents store energy in J, and
    // Buffer.Rate × dt = ΔContents lands in W·s = J as expected.
    registry["Electric Charge"] = new Resource {
      Name = "Electric Charge",
      Abbreviation = "EC",
      Unit = UnitDefinition.Watt,
      Density = 0,
      Domain = ResourceDomain.Uniform,
    };

    // Thermal energy. Storage in Joules, rate in Watts — same J/W
    // convention as EC. Heat buffers are per-device (private; never
    // added to ProcessFlowSystem) so heat from one producer can't
    // re-route into another producer's buffer through the LP's
    // proportional fill distribution.
    registry["Heat"] = new Resource {
      Name = "Heat",
      Abbreviation = "Heat",
      Unit = UnitDefinition.Watt,
      Density = 0,
      Domain = ResourceDomain.Uniform,
    };

    registry["Liquid Hydrogen"] = new Resource {
      Name = "Liquid Hydrogen",
      Abbreviation = "LH2",
      Unit = UnitDefinition.Liter,
      Density = 0.07,
      Domain = ResourceDomain.Topological,
      MliBoiloffFractionPerDay = 0.03,
      BoilingPointK = 20.3,
      LatentHeatJPerKg = 446_000,
    };

    registry["Liquid Oxygen"] = new Resource {
      Name = "Liquid Oxygen",
      Abbreviation = "LOX",
      Unit = UnitDefinition.Liter,
      Density = 1.2,
      Domain = ResourceDomain.Topological,
      MliBoiloffFractionPerDay = 0.01,
      BoilingPointK = 90.2,
      LatentHeatJPerKg = 213_000,
    };

    registry["RP-1"] = new Resource {
      Name = "RP-1",
      Abbreviation = "RP-1",
      Unit = UnitDefinition.Liter,
      Density = 0.8,
      Domain = ResourceDomain.Topological,
    };

    registry["Hydrazine"] = new Resource {
      Name = "Hydrazine",
      Abbreviation = "N2H4",
      Unit = UnitDefinition.Liter,
      Density = 1.0,
      Domain = ResourceDomain.Topological,
    };

    registry["Xenon"] = new Resource {
      Name = "Xenon",
      Abbreviation = "Xe",
      Unit = UnitDefinition.Liter,
      // KSP uses a density of around ~1 kg/L for Xenon. However, Dawn used supercritical Xenon at a
      // density of ~2 kg/L, so we use the denser value.
      Density = 2,
      Domain = ResourceDomain.Topological,
    };

    registry["Methane"] = new Resource {
      Name = "Methane",
      Abbreviation = "CH4",
      Unit = UnitDefinition.Liter,
      Density = 0.42,
      Domain = ResourceDomain.Topological,
      MliBoiloffFractionPerDay = 0.005,
      BoilingPointK = 111.7,
      LatentHeatJPerKg = 510_000,
    };

    // Nitrogen tetroxide. Hypergolic oxidizer for Hydrazine.
    registry["NTO"] = new Resource {
      Name = "NTO",
      Abbreviation = "NTO",
      Unit = UnitDefinition.Liter,
      Density = 1.45,
      Domain = ResourceDomain.Topological,
    };

    // Pre-mixed APCP-like solid grain. SRBs cast it as a single propellant
    // (no separate fuel/oxidizer plumbing), so a single resource is the
    // physically accurate model.
    registry["Solid Propellant"] = new Resource {
      Name = "Solid Propellant",
      Abbreviation = "APCP",
      Unit = UnitDefinition.Liter,
      Density = 1.8,
      Domain = ResourceDomain.Topological,
    };
  }

  public static Resource ElectricCharge => registry["Electric Charge"];
  public static Resource Heat => registry["Heat"];
  public static Resource LiquidHydrogen => registry["Liquid Hydrogen"];
  public static Resource LiquidOxygen => registry["Liquid Oxygen"];
  public static Resource RP1 => registry["RP-1"];
  public static Resource Hydrazine => registry["Hydrazine"];
  public static Resource Xenon => registry["Xenon"];
  public static Resource Methane => registry["Methane"];
  public static Resource NTO => registry["NTO"];
  public static Resource SolidPropellant => registry["Solid Propellant"];

  private Resource() {}


  public static Resource Get(string name) {
    return registry[name];
  }

  public static bool TryGet(string name, out Resource resource) {
    return registry.TryGetValue(name, out resource);
  }

  public static IEnumerable<Resource> All => registry.Values;
}

File: crates/nova-sim/src/resource.rs
//! Resource catalogue. Rust is the source of truth — KSP doesn't push
//! resource definitions across the FFI; the wire format references
//! resources by name only, and `from_name` resolves them. New
//! resources require a code change here, not a config push.
//!
//! Density is kg / unit-volume (litres for propellants, joules for
//! Electric Charge — see comment on the EC entry).

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum Resource {
    ElectricCharge,
    LiquidHydrogen,
    LiquidOxygen,
    Rp1,
    Hydrazine,
    Xenon,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum ResourceDomain {
    /// Flows along vessel topology (pipes, decouplers, drain priority).
    /// Solved by StagingFlowSystem (water-fill).
    Topological,
    /// Single vessel-wide pool — no topology distinctions, may be
    /// cyclic. Solved by ProcessFlowSystem (LP).
    Uniform,
}

impl Resource {
    /// Long form, e.g. "Liquid Hydrogen". Matches the C#
    /// `Resource.Name` registry key — used as the wire-format string.
    pub const fn name(self) -> &'static str {
        match self {
            Resource::ElectricCharge  => "Electric Charge",
            Resource::LiquidHydrogen  => "Liquid Hydrogen",
            Resource::LiquidOxygen    => "Liquid Oxygen",
            Resource::Rp1             => "RP-1",
            Resource::Hydrazine       => "Hydrazine",
            Resource::Xenon           => "Xenon",
        }
    }

    pub const fn abbreviation(self) -> &'static str {
        match self {
            Resource::ElectricCharge  => "EC",
            Resource::LiquidHydrogen  => "LH2",
            Resource::LiquidOxygen    => "LOX",
            Resource::Rp1             => "RP-1",
            Resource::Hydrazine       => "N2H4",
            Resource::Xenon           => "Xe",
        }
    }

    /// kg per unit-volume. EC has zero density (energy, not mass).
    /// Xenon uses the supercritical-storage value (~2 kg/L) to match
    /// real-world Dawn-style ion engines, not KSP's stock 1 kg/L.
    pub const fn density(self) -> f64 {
        match self {
            Resource::ElectricCharge  => 0.0,
            Resource::LiquidHydrogen  => 0.07,
            Resource::LiquidOxygen    => 1.2,
            Resource::Rp1             => 0.8,
            Resource::Hydrazine       => 1.0,
            Resource::Xenon           => 2.0,
        }
    }

    pub const fn domain(self) -> ResourceDomain {
        match self {
            Resource::ElectricCharge  => ResourceDomain::Uniform,
            Resource::LiquidHydrogen  |
            Resource::LiquidOxygen    |
            Resource::Rp1             |
            Resource::Hydrazine       |
            Resource::Xenon           => ResourceDomain::Topological,
        }
    }

    /// Resolve a wire-format resource name to its enum variant.
    /// Accepts the same long-form names as C#'s `Resource.Get(...)`.
    pub fn from_name(name: &str) -> Option<Self> {
        match name {
            "Electric Charge" => Some(Resource::ElectricCharge),
            "Liquid Hydrogen" => Some(Resource::LiquidHydrogen),
            "Liquid Oxygen"   => Some(Resource::LiquidOxygen),
            "RP-1"            => Some(Resource::Rp1),
            "Hydrazine"       => Some(Resource::Hydrazine),
            "Xenon"           => Some(Resource::Xenon),
            _                 => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const ALL: &[Resource] = &[
        Resource::ElectricCharge,
        Resource::LiquidHydrogen,
        Resource::LiquidOxygen,
        Resource::Rp1,
        Resource::Hydrazine,
        Resource::Xenon,
    ];

    #[test]
    fn name_round_trips_through_from_name() {
        for &r in ALL {
            assert_eq!(Resource::from_name(r.name()), Some(r));
        }
    }

    #[test]
    fn unknown_name_resolves_to_none() {
        assert_eq!(Resource::from_name("Unobtanium"), None);
        assert_eq!(Resource::from_name(""), None);
    }

    #[test]
    fn names_are_unique() {
        let mut names: Vec<&str> = ALL.iter().map(|r| r.name()).collect();
        names.sort();
        names.dedup();
        assert_eq!(names.len(), ALL.len());
    }

    #[test]
    fn electric_charge_is_uniform_others_are_topological() {
        assert_eq!(Resource::ElectricCharge.domain(), ResourceDomain::Uniform);
        for &r in ALL {
            if r == Resource::ElectricCharge { continue; }
            assert_eq!(r.domain(), ResourceDomain::Topological,
                       "{:?} should be Topological", r);
        }
    }

    #[test]
    fn densities_are_positive_for_propellants_zero_for_ec() {
        assert_eq!(Resource::ElectricCharge.density(), 0.0);
        for &r in ALL {
            if r == Resource::ElectricCharge { continue; }
            assert!(r.density() > 0.0, "{:?} density should be > 0", r);
        }
    }
}
